enhancements/node-healthcheck-operator.md

```diff
@@ -0,0 +1,449 @@
+
+- Provide defaults health checks and default remedy custom resource creation.
+
+- Long term - Make Machine Health Check controller only implement the external remediation api and drop the default Machine API based remediation mechanism
+
+
+### Non-Goals
+
+- Coordination of remedy operation - operators/controllers implementers
+expected to track the [node maintenance lease proposal]
+
+- Recording long-term stable history of all health-check failures or remediation
+  actions.
+
+- Replace Machine Healthcheck for IPI clusters
+
+## Proposal
+The main use-case being targeted is clusters which, for technical and/or political reasons, do not have
+the Machine-API controlling them. This enables any kind of recovery mechanism to be built and handle remediation.
+
+#### Stories
+- As an admin of a cluster I want the cluster to detect Node failures and
+ initiate recovery, so that workloads and volumes with at-most-one semantics
+  can be run/mounted elsewhere.
+
+- As an admin of a cluster I want the cluster to detect Node failures and
+ initiate recovery, so that we can attempt to restore cluster capacity.
+
+- As an OCP engineer, I want the NHC to consume the same logic for detecting
+ failures, to avoid code duplication and provide a consistent experience
+  compared to MHC.
+
+- As an OCP engineer, I want the NHC to use the same interface for initiating
+ recovery, so that we can reuse existing mechanisms (like [poison pill] and metal3) and
+  avoid code duplication.
+
+To fulfil the above we create a [Node Health Check custom resource (NCR)](anchor todo) that specifies:
+1. health criteria: list of criterias and selector
+2. template: a template of an [external custom resource (remediation CR)](link todo)
+ to create in response to a match.
+
+### Unhealthy criteria
+A node target is unhealthy when the node meets the unhealthy node conditions criteria defined.
+If any of those criterias are met for longer than given timeouts, the controller
+creates a remediation CR according to the template in the NHC CRD.
+Timeouts will be defined as part of the NHC CRD .
+
+### Implementation Details
+
+#### NodeHealthCheck CRD
+- Enable watching a node based on a label selector. 
+- Enable defining an unhealthy node criteria (based on a list of node conditions).
+- Enable setting a threshold of unhealthy nodes. If the current number is at or above this threshold no further action will take place. This can be expressed as an int or as a percentage of the total targets in the pool.
+
+E.g:
+- Create a remedy CR when node is `ready=false` or `ready=Unknown` condition for more than 5m.
+- I want to avoid triggering remediation if 40% or more of the targets of this pool are unhealthy at the same time.
+
+
+```yaml
+apiVersion: remediation.medik8s.io/v1alpha1
+kind: NodeHealthCheck
+metadata:
+  name: example
+  # this CRD is cluster-scope, namespace is not needed  
```

Comment:
Would be good to document somewhere why this should be cluster scoped

Can you think of a possible future where this is used in a centrally managed infrastructure kind of scenario, eg HyperShift where it would run on the control plane? If this is a use case you wish to support at any point, making this cluster scoped is not desirable